File: backend/services/withdrawal_service.go
package services
import (
	"errors"
	"time"
	"cronos-backend/models"
	"cronos-backend/repositories"
	"github.com/google/uuid"
)
type WithdrawalService struct {
	WithdrawalRepo *repositories.WithdrawalRepository
	UserRepo       *repositories.UserRepository
}
func NewWithdrawalService(withdrawalRepo *repositories.WithdrawalRepository, userRepo *repositories.UserRepository) *WithdrawalService {
	return &WithdrawalService{WithdrawalRepo: withdrawalRepo, UserRepo: userRepo}
}
func (s *WithdrawalService) CreateWithdrawal(userID uuid.UUID, req *models.Withdrawal) error {
	user, err := s.UserRepo.FindByID(userID)
	if err != nil {
		return errors.New("user not found")
	}
	if user.Balance < req.Amount {
		return errors.New("insufficient balance")
	}
	req.UserID = userID
	req.Status = "pending"
	if err := s.WithdrawalRepo.Create(req); err != nil {
		return errors.New("failed to create withdrawal request")
	}
	return nil
}
func (s *WithdrawalService) GetMyWithdrawals(userID uuid.UUID) ([]models.Withdrawal, error) {
	return s.WithdrawalRepo.FindByUserID(userID)
}
func (s *WithdrawalService) GetAllWithdrawals() ([]models.Withdrawal, error) {
	return s.WithdrawalRepo.FindAll()
}
func (s *WithdrawalService) UpdateWithdrawal(withdrawalID uuid.UUID, status, notes string) (*models.Withdrawal, error) {
	withdrawal, err := s.WithdrawalRepo.FindByID(withdrawalID)
	if err != nil {
		return nil, errors.New("withdrawal not found")
	}
	if withdrawal.Status != "pending" && withdrawal.Status != "approved" {
		return nil, errors.New("withdrawal cannot be updated from current status")
	}
	validStatuses := map[string]bool{"approved": true, "rejected": true, "paid": true}
	if !validStatuses[status] {
		return nil, errors.New("invalid status")
	}
	oldStatus := withdrawal.Status
	withdrawal.Status = status
	withdrawal.Notes = notes
	now := time.Now()
	withdrawal.ProcessedAt = &now
	if (status == "approved" || status == "paid") && oldStatus == "pending" {
		user, _ := s.UserRepo.FindByID(withdrawal.UserID)
		if user.Balance < withdrawal.Amount {
			return nil, errors.New("user has insufficient balance")
		}
		user.Balance -= withdrawal.Amount
		s.UserRepo.Update(user)
	}
	if status == "rejected" && oldStatus == "approved" {
		user, _ := s.UserRepo.FindByID(withdrawal.UserID)
		user.Balance += withdrawal.Amount
		s.UserRepo.Update(user)
	}
	if err := s.WithdrawalRepo.Update(withdrawal); err != nil {
		return nil, errors.New("failed to update withdrawal")
	}
	return withdrawal, nil
}

File: backend-spring/src/main/java/com/cronos/service/WithdrawalService.java
package com.cronos.service;

import com.cronos.entity.Withdrawal;
import com.cronos.entity.User;
import com.cronos.repository.WithdrawalRepository;
import com.cronos.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.*;

@Service
@RequiredArgsConstructor
public class WithdrawalService {
    private final WithdrawalRepository withdrawalRepo;
    private final UserRepository userRepo;

    public Withdrawal createWithdrawal(UUID userId, Withdrawal req) {
        User user = userRepo.findById(userId).orElseThrow(() -> new RuntimeException("user not found"));
        if (user.getBalance().compareTo(req.getAmount()) < 0) throw new RuntimeException("insufficient balance");
        req.setUserId(userId); req.setStatus("pending");
        withdrawalRepo.save(req);
        return req;
    }

    public List<Withdrawal> getMyWithdrawals(UUID userId) { return withdrawalRepo.findByUserIdOrderByCreatedAtDesc(userId); }
    public List<Withdrawal> getAllWithdrawals() { return withdrawalRepo.findAllByOrderByCreatedAtDesc(); }

    @Transactional
    public Withdrawal updateWithdrawal(UUID withdrawalId, String status, String notes) {
        Withdrawal withdrawal = withdrawalRepo.findById(withdrawalId).orElseThrow(() -> new RuntimeException("withdrawal not found"));
        if (!"pending".equals(withdrawal.getStatus()) && !"approved".equals(withdrawal.getStatus()))
            throw new RuntimeException("withdrawal cannot be updated from current status");
        if (!Set.of("approved", "rejected", "paid").contains(status)) throw new RuntimeException("invalid status");

        String oldStatus = withdrawal.getStatus();
        withdrawal.setStatus(status); withdrawal.setNotes(notes); withdrawal.setProcessedAt(LocalDateTime.now());

        if (("approved".equals(status) || "paid".equals(status)) && "pending".equals(oldStatus)) {
            User user = userRepo.findById(withdrawal.getUserId()).get();
            if (user.getBalance().compareTo(withdrawal.getAmount()) < 0) throw new RuntimeException("user has insufficient balance");
            user.setBalance(user.getBalance().subtract(withdrawal.getAmount()));
            userRepo.save(user);
        }
        if ("rejected".equals(status) && "approved".equals(oldStatus)) {
            User user = userRepo.findById(withdrawal.getUserId()).get();
            user.setBalance(user.getBalance().add(withdrawal.getAmount()));
            userRepo.save(user);
        }
        withdrawalRepo.save(withdrawal);
        return withdrawal;
    }
}
